src/System.Private.Reflection.Core/src/System/Reflection/Runtime/General/EcmaFormat/ReflectionTypeProvider.cs

```diff
@@ -0,0 +1,288 @@
+            else if (type == CommonRuntimeTypes.Byte)
+                return PrimitiveTypeCode.Byte;
+            else if (type == CommonRuntimeTypes.IntPtr)
+                return PrimitiveTypeCode.IntPtr;
+            else if (type == CommonRuntimeTypes.UIntPtr)
+                return PrimitiveTypeCode.UIntPtr;
+            else if (type == CommonRuntimeTypes.String)
+                return PrimitiveTypeCode.String;
+            else if (type == CommonRuntimeTypes.Void)
+                return PrimitiveTypeCode.Void;
+            
+            throw new ArgumentException();
+        }
+
+        public static Type GetRuntimeType(this PrimitiveTypeCode primitiveCode)
+        {
+            switch(primitiveCode)
+            {
+                case PrimitiveTypeCode.Boolean:
+                    return CommonRuntimeTypes.Boolean;
+                case PrimitiveTypeCode.Byte:
+                    return CommonRuntimeTypes.Byte;
+                case PrimitiveTypeCode.Char:
+                    return CommonRuntimeTypes.Char;
+                case PrimitiveTypeCode.Double:
+                    return CommonRuntimeTypes.Double;
+                case PrimitiveTypeCode.Int16:
+                    return CommonRuntimeTypes.Int16;
+                case PrimitiveTypeCode.Int32:
+                    return CommonRuntimeTypes.Int32;
+                case PrimitiveTypeCode.Int64:
+                    return CommonRuntimeTypes.Int64;
+                case PrimitiveTypeCode.IntPtr:
+                    return CommonRuntimeTypes.IntPtr;
+                case PrimitiveTypeCode.Object:
+                    return CommonRuntimeTypes.Object;
+                case PrimitiveTypeCode.SByte:
+                    return CommonRuntimeTypes.SByte;
+                case PrimitiveTypeCode.Single:
+                    return CommonRuntimeTypes.Single;
+                case PrimitiveTypeCode.String:
+                    return CommonRuntimeTypes.String;
+                case PrimitiveTypeCode.TypedReference:
+                    throw new PlatformNotSupportedException();
+                case PrimitiveTypeCode.UInt16:
+                    return CommonRuntimeTypes.UInt16;
+                case PrimitiveTypeCode.UInt32:
+                    return CommonRuntimeTypes.UInt32;
+                case PrimitiveTypeCode.UInt64:
+                    return CommonRuntimeTypes.UInt64;
+                case PrimitiveTypeCode.UIntPtr:
+                    return CommonRuntimeTypes.UIntPtr;
+                case PrimitiveTypeCode.Void:
+                    return CommonRuntimeTypes.Void;
+            }
+
+            throw new BadImageFormatException();
+        }
+    }
+    class ReflectionTypeProvider : ICustomAttributeTypeProvider<RuntimeTypeInfo>, ISZArrayTypeProvider<RuntimeTypeInfo>, ISignatureTypeProvider<RuntimeTypeInfo, TypeContext>
+    {
+        private static readonly RuntimeTypeInfo s_systemType = (RuntimeTypeInfo)CommonRuntimeTypes.Type.GetTypeInfo();
```

Comment:
You can get rid of both the cast and the .GetTypeInfo() call by calling the CastToRuntimeTypeInfo() extension method.